src/main/java/io/lettuce/core/masterslave/StaticMasterSlaveTopologyProvider.java

```diff
@@ -77,6 +78,13 @@ public StaticMasterSlaveTopologyProvider(RedisClient redisClient, Iterable<Redis
         Flux<RedisURI> uris = Flux.fromIterable(redisURIs);
         Mono<List<RedisNodeDescription>> nodes = uris.flatMap(uri -> getNodeDescription(connections, uri)) //
                 .collectList() //
+                .map((nodeDescriptions) -> {
+                    if (nodeDescriptions.isEmpty()) {
+                        throw new RedisConnectionException(String.format("Failed to connect to any nodes in %s", redisURIs));
```

Comment:
On second thought, this functionality could be debated on what the desired behavior is. I've asserted what _I_ think makes the most sense in the [two new tests I added](https://github.com/acmcelwee/lettuce-core/blob/434a385d4e5288b8b4ff68bef849457887a1d535/src/test/java/io/lettuce/SslTest.java#L344-L366).

If you think that it should be a hard fail on a failure to connect to _any_ node in the initial connection setup, we'd need to do something different. LMK what you think, and I'm happy to implement to continue moving this PR along.

Reply:
Thanks for raising this issue. I think you could argue in both ways, hard fail on any failure vs. fail only if no connection could be established at all. Having this safeguard in place is beneficial. Please do not use `.map(…)` but rather `flatMap(…)` and return `Mono.error(…)`. Throwing exceptions in reactive operators isn't a proper way to communicate failures.

However, this question raises another one and it became obvious why there is no SSL support for Master/Slave right now: Redis does not know about SSL (which require a tunnel/proxy), Redis topology discovery reports the actual hosts and ports. For SSL with Sentinel and Master/Slave auto-discovery we need some sort of mapper function to make it work. The reported addresses must be mapped to an URI that points to the actual SSL endpoint.

For now, I'd keep the mapper aside and introduce such a Mapper API in a new ticket.

Reply:
`.map(...)` converted to `.flatMap(...)` - 78a51f7e930a74da3144349b5012abb12fbea29b

Yeah, I fully understand why the sentinel or master/slave auto-discovery won't work as-is, and I think it'll take a bit of work to make that mapping work the way we'd need it to. I agree that it should live as a separate, isolated unit of work.

That said, I'd imagine that a lot of people will rely on the TLS support in Elasticache or Azure Cache. For us (using Elasticache), the static topology is mostly fine, but I still think it's not _quite_ perfect w/ master/slave because of how elasticache manages their endpoints. Take a cluster, say `cluster0` with two nodes, `node0` and `node1`. Elasticache basically gives me three dns endpoints along the lines of:
* `node0.cluster0.nzcsol.use1.cache.amazonaws.com` - `A` record to `node0`'s ip in one of our private vpc subnets
* `node1.cluster0.nzcsol.use1.cache.amazonaws.com` - `A` record to `node``'s ip in one of our private vpc subnets
* `master.cluster0.nzcsol.use1.cache.amazonaws.com` - `CNAME` record to either `node0` or `node1`, depending on which node is currently the primary node

Now, the problem comes in the fact that all nodes need to be specified at the time of creation of a static master/slave topology, and that the node's role (master or slave) is locked when the client is initially created. I have to specify the the `master.cluster0...` node, otherwise the client won't work in the event of a primary failover. To fully utilize all of the read replicas in the cluster, I also have to specify both of the `node0...` and `node1...` hostnames when specifying the endpoints for the static topology master/slave. That leaves me w/ a list of 3 `knownNodes` in `MasterSlaveConnectionProvider`, because whichever node is currently the master at client creation time ends up with two connections to that it, both marked as `MASTER`. As part of the dns record to reported address work, it might also be good to think about how to dedupe those connections. If we could make the static topology _slightly_ dynamic, that would be awesome. I could imagine a nice middle-ground between, dynamically discovered based on a seed node, and static_ish_ by requiring the full set of nodes listed at creation time but that will re-determine the master node at the time of failover.